lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,463 @@
+    sanitized_logging_endponts = []
+    for endpoint in logging_endpoints:
+        if isinstance(endpoint, str):
+            sanitized_logging_endponts.append(endpoint)
+        else:
+            errors.append(endpoint)
+
+    if errors:
+        if sanitized_logging_endponts:
+            logger.error(
+                f"{charm}.{logging_endpoints_getter} returned some invalid endpoint strings: {errors}"
+            )
+        else:
+            logger.error(
+                f"{charm}.{logging_endpoints_getter} should return an iterable of Loki push-api (compatible) endpoints (strings); "
+                f"got {errors} instead."
+            )
+
+    return sanitized_logging_endponts
+
+
+def _get_server_cert(server_cert_getter, self, charm):
+    if isinstance(server_cert_getter, property):
+        server_cert = server_cert_getter.__get__(self)
+    else:  # method or callable
+        server_cert = server_cert_getter(self)
+
+    if server_cert is None:
+        logger.warning(
+            f"{charm}.{server_cert_getter} returned None; sending logs over INSECURE connection."
+        )
+        return None
+    if not Path(server_cert).is_absolute():
+        raise ValueError(
+            f"{charm}.{server_cert_getter} should return a valid tls cert absolute path (string | Path)); "
+            f"got {server_cert} instead."
+        )
+    return server_cert
+
+
+def _setup_root_logger_initializer(
+    charm: Type[CharmBase],
+    logging_endpoints_getter: _GetterType,
+    server_cert_getter: Optional[_GetterType],
+    service_name: Optional[str] = None,
+):
+    """Patch the charm's initializer."""
+    original_init = charm.__init__
+
+    @functools.wraps(original_init)
+    def wrap_init(self: CharmBase, framework: Framework, *args, **kwargs):
+        original_init(self, framework, *args, **kwargs)
+
+        if not is_enabled():
+            logger.info("Charm logging DISABLED by env: skipping root logger initialization")
+            return
+
+        try:
+            logging_endpoints = _get_logging_endpoints(logging_endpoints_getter, self, charm)
+        except Exception:
+            # if anything goes wrong with retrieving the endpoint, we go on with logging disabled.
+            # better than breaking the charm.
+            logger.exception(
+                f"exception retrieving the logging "
+                f"endpoint from {charm}.{logging_endpoints_getter}; "
+                f"proceeding with charm_logging DISABLED. "
+            )
+            return
+
+        if not logging_endpoints:
+            return
+
+        juju_topology = {
+            "juju_unit": self.unit.name,
+            "juju_application": self.app.name,
+            "juju_model": self.model.name,
+            "juju_model_uuid": self.model.uuid,
+            "service_name": service_name or self.app.name,
+            "charm_type_name": type(self).__name__,
+            "dispatch_path": os.getenv("JUJU_DISPATCH_PATH", ""),
+        }
```

Comment:
Why don't we use the `JujuTopology.from_charm()` object from cosl?

Reply:
didn't want to assume the dependency or force other charms to install that. Or is it fine @simskij ?

Reply:
Use PYDEPS to make sure it's there :)